consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/statemachine/IbftController.java

```diff
@@ -58,14 +62,27 @@ public IbftController(
       final MessageTracker duplicateMessageTracker,
       final FutureMessageBuffer futureMessageBuffer,
       final SynchronizerUpdater sychronizerUpdater) {
+    this.blockchain = blockchain;
     this.ibftFinalState = ibftFinalState;
     this.ibftBlockHeightManagerFactory = ibftBlockHeightManagerFactory;
     this.futureMessageBuffer = futureMessageBuffer;
     this.gossiper = gossiper;
     this.duplicateMessageTracker = duplicateMessageTracker;
     this.sychronizerUpdater = sychronizerUpdater;
+  }
 
-    startNewHeightManager(blockchain.getChainHeadHeader());
+  public synchronized void start() {
```

Comment:
The intention is to prevent any of the active methods from running after `stop` is called.  I think that without synchronization, you could get a situation where, for example, one thread enters `handleNewBlockEvent` and continues past the `isRunning()` check, another thread is calling `stop` and moves on assuming everything is halted, but the logic in `handleNewBlockEvent` is still running.

I'm a little concerned about all the synchronization here as well -  I can take another look and see if there's a way to rework it.

Reply:
That makes sense.  Perhaps in a next pass we could use an atomic long for in-flight requests and add a lame-duck mode (finishing requests, but not accepting new ones) to the lifecycle.  It's kind of high overhead however, but right now putting sync on everything makes it all single threaded.  Perhaps documenting the reasons for synchronization in the comments would be sufficient.

Reply:
Reworked to remove the synchronization.  I have the `IbftMiningCoordinator` making sure that the `IbftProcessor` is stopped before stopping the `IbftExecutors`, which should ensure that `IbftController` doesn't try to execute anything on a stopped `IbftExecutors` instance.